tests/test_ical.py

```diff
@@ -0,0 +1,423 @@
+        self._assert_line_lengths(line_74_bytes)
+
+        # 75 octets.
+        exact_75_bytes = "x" * 72 + "あ"
+        self._assert_line_lengths(exact_75_bytes)
+
+        # 76 octets.
+        needs_adjustment = "x" * 73 + "あ"
+        self._assert_line_lengths(needs_adjustment)
+
+    def test_long_holiday(self):
+        # ASCII-only
+        ascii_fold_test = (
+            "Krungthepmahanakhon Amonrattanakosin Mahintharayutthaya Mahadilokphop "
+            "Noppharatratchathaniburirom Udomratchaniwetmahasathan Amonphimanawatansathit "
+            "Sakkathattiyawitsanukamprasit"
+        )
+        self._assert_line_lengths(ascii_fold_test)
+
+        # UTF-8
+        utf8_fold_test = (
+            "กรุงเทพมหานคร อมรรัตนโกสินทร์ มหินทรายุธยา มหาดิลกภพ นพรัตนราชธานีบูรีรมย์ "
+            "อุดมราชนิเวศน์มหาสถาน อมรพิมานอวตารสถิต สักกะทัตติยวิษณุกรรมประสิทธิ์"
+        )
+        self._assert_line_lengths(utf8_fold_test)
+
+    def test_return_bytes(self):
+        self._assert_byte_output(self.us_exporter, b"SUMMARY:New Year's Day")
+        self._assert_byte_output(self.jp_exporter, b"SUMMARY:\xe5\x85\x83\xe6\x97\xa5")
+
+    def test_crlf_line_endings(self):
+        output = self.th_exporter.generate()
+
+        lines = output.splitlines(True)
+        for line in lines:
+            self.assertTrue(
+                line.endswith(CONTENT_LINE_DELIMITER), f"Line did not end with CRLF: {repr(line)}"
+            )
+
+    def test_unique_uid_generation(self):
+        # 1st generation should yield unique UUIDs for each VEVENT.
+        output = self.us_exporter.generate()
+        uids = [line for line in output.split(CONTENT_LINE_DELIMITER) if line.startswith("UID:")]
+
+        self.assertEqual(len(uids), len(set(uids)), "Duplicate UIDs found in iCal output.")
+
+        # 2nd generation should yield different UUIDs.
+        output_2 = self.us_exporter.generate()
+        uids_2 = [
+            line for line in output_2.split(CONTENT_LINE_DELIMITER) if line.startswith("UID:")
+        ]
+
+        self.assertEqual(len(uids_2), len(set(uids_2)), "Duplicate UIDs found in 2nd iCal output.")
+
+        # Ensure that there is no overlap at all between UIDs from both attempt.
+        self.assertTrue(set(uids).isdisjoint(set(uids_2)), "Some UIDs are reused")
+
+    def test_export_ics_valid_path(self):
+        valid_path = "valid_directory"
+        os.makedirs(valid_path, exist_ok=True)
+
+        try:
+            self.us_exporter.export_ics(file_path=os.path.join(valid_path, "test_calendar.ics"))
+            self.assertTrue(
+                os.path.exists(os.path.join(valid_path, "test_calendar.ics")),
+                f"File should be created at {os.path.join(valid_path, 'test_calendar.ics')}",
+            )
+        finally:
+            os.remove(os.path.join(valid_path, "test_calendar.ics"))
+            os.rmdir(valid_path)
+
+    @patch("builtins.open", side_effect=OSError("[Errno 2] No such file or directory"))
```

Comment:
All the following tests (except for `test_export_ics_empty_content`, and possibly `test_export_ics_file_overwrite`) actually test the functionality of `os.write`, not ours. Is it necessary?

Reply:
Addressed in https://github.com/vacanza/holidays/pull/2362/commits/8240879ef4e0838aa6625864127475d6e8ac84e7 🙏